PolyEngine/RenderingDevice/OpenGL/Src/Pipeline/EnvCapture.cpp

```diff
@@ -0,0 +1,311 @@
+	gConsole.LogInfo("EnvCapture::CaptureCubemap cubemap rendered");
+
+	gConsole.LogInfo("EnvCapture::CaptureCubemap generate env cubemap mipmaps");
+	glBindTexture(GL_TEXTURE_CUBE_MAP, EnvCubemap);
+	glGenerateMipmap(GL_TEXTURE_CUBE_MAP);
+	gConsole.LogInfo("EnvCapture::CaptureCubemap env cubemap mipmaps generated");
+
+	glBindFramebuffer(GL_FRAMEBUFFER, 0);
+	glBindRenderbuffer(GL_RENDERBUFFER, 0);
+	glBindTexture(GL_TEXTURE_2D, 0);
+
+	CHECK_GL_ERR();
+}
+
+void EnvCapture::CaptureDiffuseIrradiance()
+{
+	glGenTextures(1, &IrradianceMap);
+	glBindTexture(GL_TEXTURE_CUBE_MAP, IrradianceMap);
+	for (unsigned int i = 0; i < 6; ++i)
+	{
+		glTexImage2D(GL_TEXTURE_CUBE_MAP_POSITIVE_X + i, 0, GL_RGB16F, 32, 32, 0, GL_RGB, GL_FLOAT, nullptr);
+	}
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+
+	// Create FBO for HDR capture
+	gConsole.LogInfo("EnvCapture::CaptureIrradiance Create capture FBO");
+	GLuint captureFBO;
+	GLuint captureRBO;
+	glGenFramebuffers(1, &captureFBO);
+	glGenRenderbuffers(1, &captureRBO);
+
+	glBindFramebuffer(GL_FRAMEBUFFER, captureFBO);
+	glBindRenderbuffer(GL_RENDERBUFFER, captureRBO);
+	glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, 512, 512);
+	glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, captureRBO);
+	gConsole.LogInfo("EnvCapture::CaptureIrradiance Capture FBO created");
+
+	// Projection
+	gConsole.LogInfo("EnvCapture::CaptureIrradiance create matrices");
+	Matrix uClipFromView;
+	uClipFromView.SetPerspective(90.0_deg, 1.0f, 0.1f, 10.0f);
+
+	Matrix ViewFromModel[] =
+	{
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(1.0f,  0.0f,  0.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(-1.0f,  0.0f,  0.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f, -1.0f,  0.0f), Vector(0.0f,  0.0f, -1.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  1.0f,  0.0f), Vector(0.0f,  0.0f,  1.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  0.0f,  1.0f), Vector(0.0f, -1.0f,  0.0f)),
+		Matrix(Vector(0.0f, 0.0f, 0.0f), Vector(0.0f,  0.0f, -1.0f), Vector(0.0f, -1.0f,  0.0f))
+	};
+	gConsole.LogInfo("EnvCapture::CaptureIrradiance matrices created");
+
+	glBindFramebuffer(GL_FRAMEBUFFER, captureFBO);
+	glBindRenderbuffer(GL_RENDERBUFFER, captureRBO);
+	glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, 32, 32);
+
+	CubemapIrradianceShader.BindProgram();
+	CubemapIrradianceShader.SetUniform("uEnvCubemap", 0);
+	glActiveTexture(GL_TEXTURE0);
+	glBindTexture(GL_TEXTURE_CUBE_MAP, EnvCubemap);
+
+	glViewport(0, 0, 32, 32); // don't forget to configure the viewport to the capture dimensions.
```

Comment:
You forgot here too :)

Reply:
The dimensions match size of Irradiance map (line 157).
Actually now I see that Framebuffer is bound twice and has to big depth depth attachment at first ((ines 174 and 196)